pulp_rpm/app/models/distribution.py

```diff
@@ -328,3 +331,21 @@ class Meta:
             "packages",
             "distribution_tree",
         )
+
+
+@receiver(pre_delete, sender=DistributionTree)
```

Comment:
I can't override the Model.delete method because it's not called for bulk or queryset removals.
To use pre-delete, I had to remove on_delete=PROTECT for repositories, otherwise I can remove them.
To use post_delete, I'd need to remove on_delete=CASCADE for variants and addons.

Reply:
Maybe consider using `django-lifecycle` hook. As of 3.6, all models can use this https://rsinger86.github.io/django-lifecycle/

Reply:
I'll read more about it. Does it solve the problem or order and cascade/protect rules?

Reply:
It does not, it's mainly a syntactical sugar

Reply:
Addressed in the next push

Reply:
I'm taking that back :)

with this patch:
```diff
diff --git a/pulp_rpm/app/models/distribution.py b/pulp_rpm/app/models/distribution.py
index 988ef37d..14be224c 100644
--- a/pulp_rpm/app/models/distribution.py
+++ b/pulp_rpm/app/models/distribution.py
@@ -3,8 +3,11 @@ from logging import getLogger
 
 from django.db import models
 from django.db import transaction
-from django.db.models.signals import post_delete
-from django.dispatch import receiver
+
+from django_lifecycle import (
+    AFTER_DELETE,
+    hook,
+)
 
 from pulpcore.plugin.models import (
     Artifact,
@@ -91,6 +94,24 @@ class DistributionTree(Content):
     discnum = models.IntegerField(null=True)
     totaldiscs = models.IntegerField(null=True)
 
+    @hook(AFTER_DELETE)
+    def cleanup_subrepos(self):
+        """
+        Remove subrepos when a DistributionTree is being removed.
+
+        """
+        with transaction.atomic():
+            for variant in self.variants.all():
+                subrepo = variant.repository
+                variant.delete()
+                if subrepo:
+                    subrepo.delete()
+            for addon in self.addons.all():
+                subrepo = addon.repository
+                addon.delete()
+                if subrepo:
+                    subrepo.delete()
+
     def repositories(self):
         """
         Return the subrepos in this DistributionTree.
@@ -331,22 +352,3 @@ class Variant(BaseModel):
             "packages",
             "distribution_tree",
         )
-
-
-@receiver(post_delete, sender=DistributionTree)
-def cleanup_subrepos(sender, instance, **kwargs):
-    """
-    Remove subrepos when a DistributionTree is being removed.
-
-    """
-    with transaction.atomic():
-        for variant in instance.variants.all():
-            subrepo = variant.repository
-            variant.delete()
-            if subrepo:
-                subrepo.delete()
-        for addon in instance.addons.all():
-            subrepo = addon.repository
-            addon.delete()
-            if subrepo:
-                subrepo.delete()

```

I'm getting `django.db.utils.IntegrityError: update or delete on table "rpm_distributiontree" violates foreign key constraint "rpm_addon_distribution_tree_id_91a3fbf6_fk_rpm_distr" on table "rpm_addon"
`

Reply:
I'm leaning towards keeping the signal. Django lifecycle doesn't seem to be just a syntactic sugar, it seem not to use signals at all https://github.com/rsinger86/django-lifecycle/blob/master/django_lifecycle/mixins.py#L138-L141